daemons/attrd/attrd_commands.c

```diff
@@ -955,29 +885,141 @@ attrd_peer_update(crm_node_t *peer, xmlNode *xml, const char *host, bool filter)
     if ((v->nodeid == 0) && (v->is_remote == FALSE)
         && (crm_element_value_int(xml, PCMK__XA_ATTR_NODE_ID,
                                   (int*)&v->nodeid) == 0) && (v->nodeid > 0)) {
+        record_peer_nodeid(v, host);
+    }
+}
+
+static int
+update_attr_dampening(attribute_t *a, xmlNode *xml, const char *attr)
+{
+    const char *dvalue = crm_element_value(xml, PCMK__XA_ATTR_DAMPENING);
+    int dampen = 0;
+
+    if (dvalue == NULL) {
+        crm_warn("Could not update %s: peer did not specify value for delay",
+                 attr);
+        return pcmk_rc_error;
+    }
+
+    dampen = crm_get_msec(dvalue);
+    if (dampen < 0) {
+        crm_warn("Could not update %s: invalid delay value %dms (%s)",
+                 attr, dampen, dvalue);
+        return pcmk_rc_error;
+    }
+
+    if (a->timeout_ms != dampen) {
+        mainloop_timer_del(a->timer);
+        a->timeout_ms = dampen;
+        if (dampen > 0) {
+            a->timer = mainloop_timer_add(attr, a->timeout_ms, FALSE,
+                                          attribute_timer_cb, a);
+            crm_info("Update attribute %s delay to %dms (%s)",
+                     attr, dampen, dvalue);
+        } else {
+            a->timer = NULL;
+            crm_info("Update attribute %s to remove delay", attr);
+        }
+
+        /* If dampening changed, do an immediate write-out,
+         * otherwise repeated dampening changes would prevent write-outs
+         */
+        write_or_elect_attribute(a);
+    }
+
+    return pcmk_rc_ok;
+}
+
+static void
+update_minimum_protocol_ver(const char *value)
+{
+    int ver;
+
+    pcmk__scan_min_int(value, &ver, 0);
+
+    if (ver > 0 && (minimum_protocol_version == -1 || ver < minimum_protocol_version)) {
+        minimum_protocol_version = ver;
+        crm_trace("Set minimum attrd protocol version to %d",
+                  minimum_protocol_version);
+    }
+}
+
+static attribute_t *
+populate_attribute(xmlNode *xml, const char *attr)
+{
+    attribute_t *a = NULL;
+    bool update_both = false;
+
+    const char *op = crm_element_value(xml, PCMK__XA_TASK);
+
+    if (attr == NULL) {
+        crm_warn("Could not update attribute: peer did not specify name");
+        return NULL;
+    }
```

Comment:
i think this first test is more readable in attrd_peer_update()

Reply:
Fixed.